### ISSUE TITLE
Implement proper C++ visibility support?

### ISSUE THREAD

**Issue by claudiofantacci** [robotology/superimpose-mesh-lib#15]
Implement C++ visibility support as described [here](https://gcc.gnu.org/wiki/Visibility).

This will probably allow us to remove [this Windows-specific CMake instruction](https://github.com/robotology/superimpose-mesh-lib/blob/master/CMakeLists.txt#L28).

An example of this implementation can be found in YARP in [`conf/api.h`](https://github.com/robotology/yarp/blob/master/src/libYARP_conf/include/yarp/conf/api.h.in) used, e.g., in [`os/api.h`](https://github.com/robotology/yarp/blob/master/src/libYARP_OS/include/yarp/os/api.h).

**Comment by traversaro** [comment]
The CMake's macro [`GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html) may also be useful. 

A thing that is typically tricky to get right is to correctly disable the visibility flags on Windows when building a static library. Related discussion: https://github.com/ros/console_bridge/pull/40#issuecomment-269460721 .

**Comment by claudiofantacci** [comment]
@traversaro great, thanks for your feedback. I'll have a look at it and see what I can do 👍 
I'll write here any thought I will have during the process.

**Comment by claudiofantacci** [comment]
@traversaro, the [`GenerateExportHeader`](https://cmake.org/cmake/help/v3.10/module/GenerateExportHeader.html) macro seems super cool and capable of handling the static library issue with little more effort.
I will try using it and, if succesful, include it in our [`how-to-export-cpp-library`](https://github.com/robotology/how-to-export-cpp-library) with the proper documentation 👍

**Comment by traversaro** [comment]
Yes. To be honest I think that for most cases in research code the benefit of a clean list of symbols is not worth the overhead of making sure that the visibility macros are properly used, but this is probably a decision that it make sense to do on a per-project basis.

**Comment by claudiofantacci** [comment]
> Yes. To be honest I think that for most cases in research code the benefit of a clean list of symbols is not worth the overhead of making sure that the visibility macros are properly used, but this is probably a decision that it make sense to do on a per-project basis.

100% agree 🚀

**Comment by traversaro** [comment]
The issue https://github.com/robotology/icub-main/issues/577 is a good example of issue that would be avoided if `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` was used instead of manually adding visibility decorators.
